### PR TITLE
Refactor html indent file.

### PR DESCRIPTION
**Pull request by zerustech**
Refactor the html indent file, based on the work of work of Andy Wokula. The html indent file as well as its testing script are hosted by a separate repository at [zerustch/vim-html](https://github.com/zerustech/vim-html)

### CODE REVIEW

**Comment by brammool** [comment]
Can you provide examples how this works better than the current implementation?
It drops using indent/javascript.vim, that looks suspicious.

File: runtime/indent/html.vim [hunk 1]
```diff
@@ -107,107 +350,292 @@ func! HtmlIndent_CheckUserSettings()
...
+
+  if current_block_tag_lnum > 0
+    if s:IsStartTag(current_block_tag) && current_block_tag_lnum < current_line_number
+      " It is a block start tag, and the block start tag is before current
+      " line, so a match has been found.
+      let block = get(s:indent_tags, current_block_tag)
     else
-      let b:html_indent_line_limit = 200
+      " It is a block end tag, or a block start tag in current line, try to
+      " locate an other block start tag before current line.
```
Comment (dpelle) [inline runtime/indent/html.vim]: locate an other → locate another

File: runtime/indent/html.vim [hunk 2]
```diff
@@ -244,805 +672,1304 @@ call s:AddITags(s:indent_tags, [
...
-  if state.lnum == 0
-    return state
+  let end_tag = matchstr(a:text, s:StartsWithEndTagPattern())
+
+  " The string starts with a normal end tag; and
+  " the root block is either 0 (no root block) or 6 (conditional comment); and
+  " the root block has not been reset by any block end tags.
+  if end_tag != '' && (b:tag_counter.root_block == 0 || b:tag_counter.root_block == 6) && 'guard' == '' . b:tag_counter.block_stack[0]
+    " It is not really necessary to update current_line_offset_stack here,
+    " because it will not be used any more. However, it is no harmful to
```
Comment (dpelle) [inline runtime/indent/html.vim]: it is no harmful → it is not harmful

File: runtime/indent/html.vim [hunk 3]
```diff
@@ -244,805 +672,1304 @@ call s:AddITags(s:indent_tags, [
...
+  if len(stack) > 0 | let stack = reverse(stack) | endif
+
+  for synid in stack
+    let name = synIDattr(synid, "name")
+    if index(b:hi_insideTagNames, name) >= 0
+      let context.ready = 1
+      let context.state = 'line_inside_tag'
+      " When calcuating indent of an attribute, the b:context.context is not
+      " used, so there is no need to update context.indent here.
+      " But it is no harmful to set this value to keep consistent logic and
```
Comment (dpelle) [inline runtime/indent/html.vim]: it is no harmful → it is not harmful

File: runtime/indent/html.vim [hunk 4]
```diff
@@ -244,805 +672,1304 @@ call s:AddITags(s:indent_tags, [
...
+      " ... <- This is current line, it should be aligned as a normal line, so
+      " context.ready SHOULD NOT be set to 1
+      " <![endif]-->
     else
-      let text = tolower(getline(comlnum)[: comcol-2])
+      let context.ready = 1
+
+      " Set the fallback indent of current line. Actually, it is not necessary,
+      " because lines in block are to be indented by Alien{*} methods. However,
+      " it is no harmful to set context.indent here to keep consistent logic
```
Comment (dpelle) [inline runtime/indent/html.vim]: it is no harmful → it is not harmful

File: runtime/indent/html.vim [hunk 5]
```diff
@@ -244,805 +672,1304 @@ call s:AddITags(s:indent_tags, [
...
+  " Current line does not start with any end tag, try to initialize context as
+  " if its previous line ends with an end tag.
+  let context = s:InitContextOfLineAfterEndTag(context)
+  if context.ready | return context | endif
+
+  " The previous line does not end with any end tag, try to initialize context
+  " as if it is after a block.
+  " We should check if current line starts with an end tag, or if its previous
+  " line ends with an end tag before this method, because a line after a block
+  " may starts with or be after an end tag as well.
```
Comment (dpelle) [inline runtime/indent/html.vim]: may starts → may start

File: runtime/indent/html.vim [hunk 6]
```diff
@@ -244,805 +672,1304 @@ call s:AddITags(s:indent_tags, [
...
-      endif
-    else
-      " block continues
-      " indent this line with alien method
-      let indent = s:Alien{b:hi_indent.block}()
-    endif
-  else
-    " not within a block - within usual html
-    let b:hi_newstate.block = b:hi_indent.block
-
```
Comment (dpelle) [inline runtime/indent/html.vim]: it may starts → it may start

**Comment by zerustech** [comment]
@brammool Thanks for your feedback. I have revised the code base so that indent/javascript.vim and syntax/javascript.vim can be reused for indenting JavaScript code. And please see my notes below about the improvements of this script:

- Fixed some bugs ( I didn't keep track of all the issues I fixed, because I have refactored almost everything)
- Reuse indent/css.vim and syntax/css.vim for indenting CSS code.
- Reuse indent/javascript.vim and syntax/javascript.vim for indenting JavaScript code.
- Support custom block tags and indent methods (i.e., '{#', and '#}' in twig).
- Generate nice looking three-piece comments (including a workaround for issue #1696) for HTML, Javascript, and CSS.
- Support non-standard tags that do not start with '<', or end with '>', by using meta data instead of hard coding.
- Refactored the code base, and documented the algorithm with detailed comments, to make future maintenance much easier.
- Full unit test script, based on [vader.vim](https://github.com/junegunn/vader.vim), has been developed and hosted at [zerustech/vim-html](https://github.com/zerustech/vim-html)

**Comment by zerustech** [comment]
@dpelle Thanks for your feedback. The typos have been fixed.

**Comment by keelii** [comment]
This PR actually solve my problem
style_in_html_indent_broken

**Comment by jonsmithers** [comment]
@keelii see `:help html-indenting`. That issue should go away if you `:let g:html_indent_style1 = "inc"`. It took me forever to figure that out. It seems to me that it'd be better if it defaulted to `"inc"` instead of `"zero"`.

**Comment by keelii** [comment]
@jonsmithers That looks nice, thank you

**Comment by chrisbra** [comment]
Current maintainer is Bram, so please correct me if I am wrong here @brammool . I cannot speak for him, but I know he is busy with developing and enhancing the Vim source, even without having to handle all the runtime files. So I bet I would be a relief for him, to take off some of his responsibility. So to make it long story short, would you be willing to maintain the indent file, since you already did a major refactor here?

**Comment by brammool** [comment]
We now have support for testing indent files.  It would be good if you can show your improvements with a test (fails with old code, succeeds with your new version).  I know it's going to be some work, but it's good to have tests to avoid regressing later.

**Comment by brammool** [comment]
Looks like this might still be useful, especially if we can add tests for HTML indenting.  Ideally we first add tests that also pass with the existing indent file, than use the improved indenting that will show fixing test cases that failed before.

**Comment by chrisbra** [comment]
@zerustech can you please rebase? We have indent tests in place, can you please also add a few new ones (preferably some that do not work well currently)

**Comment by chrisbra** [comment]
closing for now

**Pull request closed without merging**